### DISCUSSION TITLE
The Core Algorithmn does not work anymore! Please read this!

### DISCUSSION THREAD

**Discussion by drbacke** [Akkudoktor-EOS/EOS#137] [Announcements]
I'm afraid that I may not be able to save the code. So many changes have been made to the two files (class_ems / class_optimize), both files are very complex. 
In particular, it is very important to pay attention to the times, hour start status and end status. Parts of the calculation are simply gone. 
The JSON implementation is also completly broken. 
I would urge all developers not to change any more files from the core algorithm. You can't optimize anything without having understood it!
At least until all unit tests are sufficiently good and we have a halfway decent documentation / API documentation. A Docker release would also be nice to make it easier to install.
I will take some time over the next few days and try to remove all the bugs. I may have to revert, though. 
Hopefully not

**Comment by michaelosthege** [comment]
@drbacke yes, PRs have been merged too fast.

I would suggest to invest half an hour into bringing everybody who's contributing maintenance work onto the same page, also with respect to how things work and why.
It will be much more worthwhile than doubling down by one person followed by seveal people trying to rebase PRs to a main branch that has changed dramatically.

#126 is coming to fix the Docker setup

#128 cleans up the test for the optimization parts to the point where the test is easier to work with and extend. (Please take a look at this)

**Comment by b0661** [comment]
@drbacke, I appreciate going on with a more test driven development.

As you merged the server test #135, can you please also merge #125. Even if you plan to make a cleanup in this area. #125 only adds back load_profiles.npz to the data directory.

Without #125 all tests will fail (rightly, because load_profiles.npz is missing).

**Comment by drbacke** [comment]
Sorry but the code is entirely destroyed. The whole JSON format is destroyed, the algorithm returns 9000 all the time. There are so many bugs in the code, I dont know where to start.
The PRs make it even worse.  
I have to correct everything. This will take weeks. 
I should check who did this and kick him out

**Reply by b0661** [reply]
I see the problem. Especially if everything seems to be destroyed a working test environment is key. That´s why I would really appreciated PR #125 to be merged. This allows testing without constant unrelated error messages.

Maybe you can identify some "hotspot" files where we can write tests taking the original version as a starting point. Git never forgets!

**Reply by NormannK** [reply]
[test.zip](https://github.com/user-attachments/files/17310022/test.zip)
also we had merges in the past where checks failed and some who had open questions.

I think the problem is in the single_test_optimization.py itself. I did replace all modules with very old versions and still had the problem. Only changing the single_test_optimization.py to an old version did solve it. Ill continue looking into it.

Edit: using the current code and adding to an old test version '"min_soc_prozent": 15"' as a parameter does produce the correct output. So I suspect PR #113

Attached an old test.py with "min_soc_prozent": 15",  verbose=True, fixed_seed=42. No other changes

**Reply by NormannK** [reply]
@drbacke found the error. 
with PR #113 you changed eauto_min_soc to 1000. That was 0 before.

**Reply by drbacke** [reply]
I fixed some problems, next one will be the wrong timeshift in the results.
